### PR TITLE
Migrate from Policy MergeOptions to ToFieldPath in `beta convert pipeline-composition`

### LINKED ISSUE #5421: `beta convert pipeline-composition` doesn't handle `mergeOptions`
### What happened?

I upgraded crossplane to `1.15` (from `1.14.x`)
I transformed Composition to use patch-and-transform functions via `crossplane beta convert pipeline-composition`

In Composition I am trying to append list of strings to  

```
        # adding ns specific SG which allow connectivity to app from DataCenter or AWS EC2 Instance
        - fromFieldPath: metadata.annotations[mysqlinstance.crossplane.sinch.com/namespaceSG]
          policy:
            mergeOptions:
              appendSlice: true
          toFieldPath: spec.forProvider.vpcSecurityGroupIds[0]
          type: FromCompositeFieldPath
         # adding environment specific SG (e.g. which allow connectivity from BastionHost)
        - fromFieldPath: mysqlinstance.envConfig.additionalSecurityGroups
          policy:
            mergeOptions:
              appendSlice: true
          toFieldPath: spec.forProvider.vpcSecurityGroupIds
          type: FromEnvironmentFieldPath
```

Before upgrading crossplane and changing composition to Pipeline mode (Patch And Transform) everything was working as expected, after above operations second patch is overriding first element in an array set by previous patch

Expected behaviour: 
It should append to the existing list new values (instead replacing previous elements)

### How can we reproduce it?

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  labels:
    crossplane.io/xrd: xmysqlinstances.rds.aws.crossplane.sinch.com
  name: mysqlinstances.rds.aws.crossplane.sinch.com-v1alpha1
spec:
  compositeTypeRef:
    apiVersion: rds.aws.crossplane.sinch.com/v1alpha1
    kind: XMySQLInstance
  environment:
    environmentConfigs:
    - ref:
        name: mysqlinstance
      type: Reference
  mode: Pipeline
  pipeline:
  - functionRef:
      name: function-patch-and-transform
    input:
      apiVersion: pt.fn.crossplane.io/v1beta1
      environment: null
      kind: Resources
      patchSets: []
      resources:
      - base:
          apiVersion: database.aws.crossplane.io/v1beta1
          kind: RDSInstance
          spec:
            forProvider:
              caCertificateIdentifier: rds-ca-rsa2048-g1
              engine: mysql
              masterUsername: root
              publiclyAccessible: false
            managementPolicies:
            - '*'
            providerConfigRef:
              name: crossplane-provider-aws
            publishConnectionDetailsTo:
              configRef:
                name: default
        name: rds
        patches:
        # adding ns specific SG which allow connectivity to app from DataCenter or AWS EC2 Instance
        - fromFieldPath: metadata.annotations[mysqlinstance.crossplane.sinch.com/namespaceSG]
          policy:
            mergeOptions:
              appendSlice: true
          toFieldPath: spec.forProvider.vpcSecurityGroupIds[0]
          type: FromCompositeFieldPath
        # adding environment specific SG (e.g. which allow connectivity from BastionHost)
        - fromFieldPath: mysqlinstance.envConfig.additionalSecurityGroups
          policy:
            mergeOptions:
              appendSlice: true
          toFieldPath: spec.forProvider.vpcSecurityGroupIds
          type: FromEnvironmentFieldPath
    ....
---
apiVersion: apiextensions.crossplane.io/v1alpha1
data:
  mysqlinstance:
    envConfig:
      additionalSecurityGroups:
      - sg-06d7ecbe89aabc75a
      - sg-048580a25883584ad
kind: EnvironmentConfig
metadata:
  name: mysqlinstance
---
apiVersion: rds.aws.crossplane.sinch.com/v1alpha1
kind: MySQLInstance
metadata:
  annotations:
    mysqlinstance.crossplane.sinch.com/namespaceSG: sg-xxx
  name: mysql-7
  namespace: ns-with-custom-sg
spec:
  parameters: {}
```

### What environment did it happen in?

crossplane: `v1.15.0`
aws-provider: `crossplane-contrib/provider-aws:v0.46.0`
kubernetes: `1.26.x`
Kubernetes distribution: `local/kind`

### PR DESCRIPTION
**Pull request by padlar**
### Description of your changes

Fixes #5421 

ℹ️  **Key implementation note**: The types in this PR follow a consistent pattern of a struct that embeds the underlying `Patch` type (that contains the old policy format) while also inserting a new `Policy` field at the same level of this embedded `Patch` struct. When the converted data is serialized to the final converted output manifest, this has the effect of essentially overriding the deprecated policy format with our new policy format.

### Useful resources for reviewers:

* Definition of what function-p-and-t supports and how it differs: https://github.com/crossplane-contrib/function-patch-and-transform/?tab=readme-ov-file#differences-from-the-native-implementation
* @negz suggested design for this implementation: https://github.com/crossplane-contrib/function-patch-and-transform/issues/101#issuecomment-2024155035
* PR that updated function-p-and-t: https://github.com/crossplane-contrib/function-patch-and-transform/pull/102/files

Here is a logic table that captures the conversion rules we need to codify in this function. I've kept the case ordering @negz used in his linked comments above for consistency:

| # | `MergeOptions` | `appendSlice`   | `keepMapValues` | `policy.toFieldPath` |
| -----------  | -----------  | ------------- | -----------  | ------------- |
| 1 | `nil`  | N/A | N/A | `nil` which defaults to `Replace` |
| 2 | `non-nil` | `nil` or `false` | `true` | `MergeObjects` | 
| 3 | `non-nil` | `true` | `nil` or `false` | `ForceMergeObjectsAppendArrays` |
| 4 | `non-nil` | `nil` or `false` |  `nil` or `false` |  `ForceMergeObjects`   |
| 5 | `non-nil` | `true` | `true` | `MergeObjectsAppendArrays` |

### Manual testing

A gist containing test manifests and expected output can be found at https://gist.github.com/jbw976/1c06224c0edc4cf9ca9a3765539a1e7f

In its source composition file, it contains patches for all the possible `mergeOptions` combinations in all the possible locations (environment, patchsets, resources). The converted composition has been verified to contain properly converted policies in their correct locations. As a further validation step, the converted composition has been used with the `render` command to sanity test its structure and content.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] ~Added or updated e2e tests.~
- [x] Linked a PR or a [docs tracking issue] to [document this change].
  - https://github.com/crossplane-contrib/function-patch-and-transform/pull/112 
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

**Comment by padlar** [comment]
@phisco I took a stab at migrating away from mergeoptions https://github.com/crossplane/crossplane/issues/5421, not sure if I am on the right track?

File: apis/apiextensions/v1/composition_patches.go
Comment (negz) [inline apis/apiextensions/v1/composition_patches.go:63]: We can't delete `MergeOptions` here in core Crossplane. It would be a breaking API change.

We could introduce `ToFieldPath`, support both, and add a comment to `MergeOptions` saying that it's deprecated.

Comment (padlar) [reply apis/apiextensions/v1/composition_patches.go:63]: Regarding last line, just wondering how is it possible to replace MergeOptions with ToFieldPath without touching core? Since MergeOptions is a field inside PatchPolicy https://github.com/crossplane/crossplane/blob/b565aaccf9a9a2e87a20c693d476b8be81bd6994/apis/apiextensions/v1/composition_patches.go#L114

Comment (negz) [reply apis/apiextensions/v1/composition_patches.go:63]: See https://github.com/crossplane/crossplane/issues/5421#issuecomment-1985273029.

If you're trying to fix #5421, the point of that issue is to update tooling that people can use to migrate from native patch and transform to https://github.com/crossplane-contrib/function-patch-and-transform. The former uses `MergeOptions`, the latter uses `ToFieldPath` instead.

What you're doing here is instead changing the native P&T implementation to use `ToFieldPath` instead of `MergeOptions`. 

(P.S. Please follow the instructions in the PR template. Especially the part about using Fixes to make it clear what issue you're trying to fix.)

Comment (padlar) [reply apis/apiextensions/v1/composition_patches.go]: @negz Thanks for the hint! I have limited my changes to converter.go. Its still in draft as I hnvt updated the tests.

**Comment by negz** [comment]
@padlar Heads up, we realized we might have got this new API wrong. 😬 Please see https://github.com/crossplane-contrib/function-patch-and-transform/issues/101 for details. Please hold off on this PR until we update function-patch-and-transform.

**Comment by jbw976** [comment]
@padlar thanks for your patience! looks like https://github.com/crossplane-contrib/function-patch-and-transform/issues/101 has been resolved now with https://github.com/crossplane-contrib/function-patch-and-transform/pull/105, so we should be able to proceed again on this work in the CLI to update the `convert` command.

It would be great to get that included for v1.16 - code freeze is 1 week from tomorrow (Tues Apr 30). I'll check back with you soon here, but let me know if any assistance is needed. Thank you!! 🙇‍♂️

File: apis/apiextensions/v1/composition_patches.go
Comment (padlar) [reply apis/apiextensions/v1/composition_patches.go]: @negz I have now updated the converter file to migrate to ToFieldPath. I had to duplicate some of the structs to achieve this - let me know wdyt? I will fix the e2e tests and add couple of unit tests to cover different migration scenarios :)

**Comment by padlar** [comment]
Thanks for the update @jbw976. I made some changes to the code yesterday which fixes the issue but I'm not 100% sure if that's the best approach to solving it. I have pinged @negz for a quick review.

**Comment by jbw976** [comment]
Awesome @padlar, thanks for the quick turnaround on updating this PR!! I'll try to take a deeper look myself today too, but in the meantime, can you share some testing information/resources for this PR? e.g. update the PR description to include the manifests you are testing with, and how you have verified they are correct. 🙇‍♂️ 

Sometimes I do this with a github gist, which can be useful for sharing the full steps and resources in one place. Here's a recent example of mine: https://gist.github.com/jbw976/4d145d97c1de7f3fa57a7bca437bbe4b

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (jbw976) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:237]: Let's move these types into [`types.go`](https://github.com/crossplane/crossplane/blob/master/cmd/crank/beta/convert/pipelinecomposition/types.go) so they are all discoverable in one place.

**Review by jbw976 — COMMENTED** [review]
Thanks for continuing to make progress on this @padlar, it's really appreciated!

I think in general, this approach is reasonable. I was hoping we could import the types from function-patch-and-transform, but I see from https://github.com/crossplane/crossplane/blob/master/cmd/crank/beta/convert/pipelinecomposition/types.go#L21-L23 that isn't possible.  We've already established a pattern of copying types from function-p-and-t to here, so this approach already has precedent.

That said, here are some considerations for you:

* Put all the copied types into `types.go` for easy discoverability in one place
* I doubt we need to give them new names, e.g. `NewPatchSet`, I expect we can just keep their names as is because they are in this package, separate from the typical `apiextensions.crossplane.io/v1` types.
* ⚠️ We need to be careful when we do copy these types though. Previously, this `convert` command has relied on using mostly the `apiextensions.crossplane.io/v1` types which will serialize out to compatible types in function-p-and-t. When we copy a type from function-p-and-t to here, but don't get it 100% correct, we risk regressing functionality that has already been working OK. So we have to be careful 😉 

Let's see how this looks with the new unit test to cover the core logic of `patchPolicy()` as well as sharing the manifests you've been testing with. It would be great to build confidence that we're getting this right! Let me know any way I can assist, I'll be sure to test your updates on this PR and help drive confidence with you 💪

**Comment by padlar** [comment]
Tested with the following composition files 
[composition.txt](https://github.com/crossplane/crossplane/files/15106832/composition.txt)
[converted-composition.txt](https://github.com/crossplane/crossplane/files/15106833/converted-composition.txt)

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: Moved to types.go

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: skipped

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: yes refactored the common code into a new function

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go:207]: I have updated the code to cover new cases and add unit tests to test patchPolicy

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (jbw976) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:216]: I don't think this logic is 100% correct for the conversion rules we want to enable. Let's review the following resources to guide the conversion logic this function should capture:

* https://github.com/crossplane-contrib/function-patch-and-transform/issues/101#issuecomment-2024141807
* https://github.com/crossplane-contrib/function-patch-and-transform/issues/101#issuecomment-2024155035
* https://github.com/crossplane-contrib/function-patch-and-transform/?tab=readme-ov-file#differences-from-the-native-implementation


Issues:
* `fromFieldPath` probably doesn't matter, we should be able to ignore that, right?
* it doesn't handle the case of `Replace`
* it is only considering values being `nil`, it's not handling values being set explicitly to `false`

Here is a logic table that captures the conversion rules we need to codify in this function. I've kept the case ordering @negz used in his linked comments above for consistency:

| # | `MergeOptions` | `appendSlice`   | `keepMapValues` | `policy.toFieldPath` |
| -----------  | -----------  | ------------- | -----------  | ------------- |
| 1 | `nil`  | N/A | N/A | `Replace` |
| 2 | `non-nil` | `nil` or `false` | `true` | `MergeObjects` | 
| 3 | `non-nil` | `true` | `nil` or `false` | `ForceMergeObjectsAppendArrays` |
| 4 | `non-nil` | `nil` or `false` |  `nil` or `false` |  `ForceMergeObjects`   |
| 5 | `non-nil` | `true` | `true` | `MergeObjectsAppendArrays` |

Because this function is only considering `nil` values, the results it produces doesn't quite match up with what I expect. For example, if the function takes in `{appendSlice: true, keepMapValues: false}` (case 3 above), it produces the incorrect result of `MergeObjectsAppendArrays`.

What do you think of this analysis, is it missing anything? If you agree with it, then please update this function to cover these cases and produce these results 🙏

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (jbw976) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:208]: I don't think `FromFieldPath` would matter at all. The way this is written, the function would return early and not produce any policy result if `FromFieldPath` is `nil` but there are valid `MergeOptions` that we should be processing.

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (jbw976) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:241]: I think it's reasonable to include these in `types.go` also, they are part of the type information.

**Review by jbw976 — CHANGES_REQUESTED** [review]
Awesome @padlar, thanks for incorporating the feedback so far. The testing information is also really helpful! The structure of this PR does look pretty much like I expected it to be now, but there are some questions I have about the functional correctness of it. 

Can you take a close look at my feedback on the core logic and see if you agree with it? Thank you very much! 🙇‍♂️

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go:212]: Thanks for the detailed summary on this, I have handled all the cases now.

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: yes totally agree, fixed now :)

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: good catch, moved!

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (jbw976) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go:212]: Awesome @padlar, the functionality does indeed look correct now, and the new unit tests are thorough with good coverage for the cases! Nice work, thank you for getting this functionality to something we can rely on 😉 

One thing I realized is that it's probably better to just set the converted policy to `nil` when nothing was specified in the original manifest, so we aren't adding config (bloat) that wasn't actually in the original manifest. That means that case 1 above should just be `nil` meaning function-p-and-t will use its default of `Replace` ([code link](https://github.com/crossplane-contrib/function-patch-and-transform/blob/main/input/v1beta1/resources_patches.go#L90)). I edited the above table to reflect that.

Another thing I was thinking of is the readability and future maintainability of this logic. We took awhile together to actually figure out what the logic is supposed to be, so I was thinking it would be wise to make this function super readable for our future selves to look at it and immediately know what it's doing.

I took a stab at a very readable approach in https://github.com/jbw976/crossplane/tree/merge-options-conversion, specifically https://github.com/jbw976/crossplane/commit/e56e06678e664d3e2b2c517a79605a5d81559804. What do you think of that style? I think I like it because it feels more natural and quickly readable. What do you think about adopting it? Feel free to just copy that code, no need to add me as a committer or anything 🤓

Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: Good catch on setting the converting policy to `nil`. And totally agree on making the conversion logic simple to follow and maintain :) I have copy pasted your code now :)

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (jbw976) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:209]: and this one could be `migratePatchPolicy`

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (jbw976) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:224]: and this one could be `migrateMergeOptions`

**Review by jbw976 — COMMENTED** [review]
Awesome @padlar, this is looking very good to me now! I took it for a full test run as well, using the manifests in https://gist.github.com/jbw976/1c06224c0edc4cf9ca9a3765539a1e7f and everything does look to be working correctly, for both `patchSets` and `resources` 💪 🤩 

I've just left a few suggestions for cleaning up a bit - mainly naming and comments/context, but nothing functional.  If you want to address those, then I think we can approve this one! 🙇‍♂️

**Comment by jbw976** [comment]
We'll likely be cutting the release branch today @padlar as per Code Freeze in https://github.com/crossplane/release/issues/9 - do you think you'll be able to update this PR with the final feedback today so we can hopefully get it merged? 🤩

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: updated

File: cmd/crank/beta/convert/pipelinecomposition/types.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/types.go]: updated

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go:205]: added comment

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: added comment

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: updated

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: updated

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go:172]: added a comment

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: updated the description

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (padlar) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: updated

**Comment by padlar** [comment]
Addressed all the feedback. If the pipeline turns green, it should be good to go :)

**Review by jbw976 — APPROVED** [review]
Great @padlar, this looks good to me!! Thanks for the continued effort you invested into this PR to get it here, really appreciate your time and effort on this important work! 🙇‍♂️

File: cmd/crank/beta/convert/pipelinecomposition/converter.go
Comment (phisco) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:151]: we should also migrate patches in the environment 😅

Comment (jbw976) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go:151]: ah, good catch @phisco! 🙇‍♂️ 

@padlar I can add a commit to this PR to handle the environment patches too 😉

Comment (jbw976) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go]: @phisco @padlar I pushed a new commit that implements migration for environment patches too. It also added more unit test coverage for all the `MigratePatchPoliciesIn*` functions.

Please take a look when you can! 🙇‍♂️

**Review by phisco — APPROVED** [review]
Thanks @padlar and @jbw976! 🙏

**Comment by jbw976** [comment]
Thanks for the review/approval @phisco! feeling good about these changes 💪 

@padlar for full transparency, I rebased everything from your branch onto the latest in master last night, then pushed my commit for environment patches and full unit tests this morning. When I did that, I believe I overwrote your commit to handle environment patches (https://github.com/crossplane/crossplane/commit/a0609bdf286b5e953328e705a85d3e3299ca579b), so I sincerely apologize for that oversight! 

So we've got your first 9 commits for everything except environment patches, then we've got my 1 commit for environment patches and unit tests. We've reviewed all that, the unit tests are passing, we've manually tested it - it all looks good.  Let's go ahead and merge that to get the feature in for the release. You'll still have your 9 authored commits in the history/stats, but once again sorry for the oversight where we did simultaneous work that conflicted with each other 🙇‍♂️

**Pull request merged**
